subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -352,183 +274,159 @@ static void device_twin_request(void)
+		.data.msg.payload = payload,
+		.topic.ptr = topic.ptr,
+		.topic.size = topic.size,
+		.topic.properties = properties,
 	};
+	az_span topic_span = az_span_create(topic.ptr, topic.size);
+	az_iot_hub_client_c2d_request c2d_msg;
+	az_iot_hub_client_method_request method_request;
+	az_iot_hub_client_twin_response twin_msg;
 
-	LOG_DBG("MQTT_EVT_PUBLISH: id = %d, len = %d ",
-		p->message_id, payload_len);
-
-	err = publish_get_payload(client, payload_len);
-	if (err) {
-		LOG_ERR("publish_get_payload, error: %d", err);
+	/* The topic needs to be parsed with multiple functions in order
+	 * to determine the type of incoming message.
+	 */
+	if (az_result_succeeded(
+	    az_iot_hub_client_c2d_parse_received_topic(&iot_hub_client, topic_span, &c2d_msg))) {
+		/* The message was received on the device-bound (C2D) topic. */
+		evt.topic.type = AZURE_IOT_HUB_TOPIC_DEVICEBOUND;
+		LOG_DBG("Message type: Device-bound");
+	} else if (az_result_succeeded(
+		   az_iot_hub_client_methods_parse_received_topic(&iot_hub_client,
+								  topic_span,
+								  &method_request))) {
+		/* The message was received on the device method topic. */
+		evt.topic.type = AZURE_IOT_HUB_TOPIC_DIRECT_METHOD;
+		LOG_DBG("Message type: Direct method");
+	} else if (az_result_succeeded(
+		   az_iot_hub_client_twin_parse_received_topic(&iot_hub_client,
+							       topic_span,
+							       &twin_msg))) {
+		/* Message was received on the device twin topic.
+		 * Now we need to determine what type of device twin message it is.
+		 */
+		switch (twin_msg.response_type) {
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_GET:
+			evt.topic.type = AZURE_IOT_HUB_TOPIC_TWIN_DESIRED;
+			LOG_DBG("Message type: AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_GET");
+			break;
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES:
+			evt.topic.type = AZURE_IOT_HUB_TOPIC_TWIN_DESIRED;
+			LOG_DBG("Message type: "
+				"AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES");
+			break;
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_REPORTED_PROPERTIES:
+			evt.topic.type = AZURE_IOT_HUB_TOPIC_TWIN_REQUEST_RESULT;
+			LOG_DBG("Message type: "
+				"AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_REPORTED_PROPERTIES");
+			break;
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_REQUEST_ERROR:
+			evt.topic.type = AZURE_IOT_HUB_TOPIC_TWIN_REQUEST_RESULT;
+			LOG_DBG("Message type: "
+				"AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_REQUEST_ERROR");
+			break;
+		default:
+			LOG_ERR("Unrecognized twin message type: %d. Treated as C2D message",
+				twin_msg.response_type);
+			evt.topic.type = AZURE_IOT_HUB_TOPIC_DEVICEBOUND;
+			break;
+		}
+	} else {
+		LOG_ERR("Unrecognized topic: %.*s", topic.size, topic.ptr);
+		evt.topic.type = AZURE_IOT_HUB_TOPIC_UNKNOWN;
 		return;
 	}
 
-	if (p->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
-		const struct mqtt_puback_param ack = {
-			.message_id = p->message_id
-		};
+	switch (evt.topic.type) {
+	case AZURE_IOT_HUB_TOPIC_DEVICEBOUND: {
```

Comment:
This switch-case (AZURE_IOT_HUB_TOPIC_DEVICEBOUND) is almost 30 lines long, an it looks like it can be extraced into it's own function without too many arguments.